lib/utils.nix

```diff
@@ -125,27 +125,185 @@ rec {
     end
   '';
 
+  /**
+    Generic "heredoc-style" escape utility.
+
+    # Args
+
+    - string: The string to escape.
+    - buildSurround: A function that maps Integer to Surround. Described in detail below.
+    - surroundWithNewlines: A boolean that impacts how to surround the string when escaping it.
+
+    A Surround is an attr set that looks like this:
+
+    ```nix
+    {
+      open = "(";
+      close = ")";
+    }
+    ```
+
+    `buildSurround` is a function that takes a depth Integer and returns a
+    Surround. For exmple, to build the previous Surround:
+
+    ```nix
+    nix-repl> buildSurround = depth: {
+      open = lib.strings.replicate depth "(";
+      close = lib.strings.replicate depth ")";
+    }
+    nix-repl> buildSurround 1
+    {
+      close = ")";
+      open = "(";
+    }
+    ```
+
+    We invoke `buildSurround` builder starting a depth 0, and increasing the
+    depth until we have a Surround where `surround.close` is not in the
+    string we're escaping.
+
+    We return the original string, first (possibly, depending on
+    `surroundWithNewlines`) surrounded by newlines , and secondly surrounded by
+    the Surround we found above.
+    ```
+  */
+  escapeString =
+    {
+      string,
+      buildSurround,
+      surroundWithNewlines,
+    }:
+    let
+      findSurround =
+        {
+          string,
+          buildSurround,
+          depth,
+        }:
+        let
+          surround = buildSurround depth;
+        in
+        if lib.hasInfix surround.close string then
+          findSurround {
+            inherit string buildSurround;
+            depth = depth + 1;
+          }
+        else
+          surround;
+
+      surround = findSurround {
+        inherit string buildSurround;
+        depth = 0;
+      };
+      newlines = if surroundWithNewlines then "\n" else "";
+    in
+    "${surround.open}${newlines}${string}${newlines}${surround.close}";
```

Comment:
I wonder if this is more readable as string concatenation instead of interpolation? Unsure.